### ISSUE TITLE
Missed breakpoints when files imported in bin/ folder use relative paths (not package: uris)

### ISSUE THREAD

**Issue by DanTup** [Dart-Code/Dart-Code#1446]
Splitting this from #1443 since that one is not related to `package:` URIs but part/part-of!

---

@DanTup:

This seems to be caused by relative path imports from outside lib. Since the entry point is outside `lib` it isn't considered a `package:` uri, and therefore the relative path is also not a `package:uri`.

A quick workaround is to change your imports in your `bin/` scripts to be `package:cfa/xxx` imports (I think this is probably the preferred way to write them anyway?).

@lrhn @peter-ahe-google before I start multiplying up breakpoints again, is there a better for for this in Dart? For example, could/should imports into a packages `lib` folder also be recognised and mapped to `package:` uris (for the same reasons #33076 was done)?

---

@peter-ahe-google:

I think we can do something similar to dart-lang/sdk#33076. It requires a minor tweak to the language specification, and @lrhn is running that by the rest of the language team.

**Comment by mezoni** [comment]
It seems that there is no activity in [https://github.com/dart-lang/sdk/issues/35860](https://github.com/dart-lang/sdk/issues/35860).
It would be interesting to know in other development environments is this also a problem of debugging?
For example, in JetBrains IDEs.

**Comment by DanTup** [comment]
It's possible that if they send the breakpoints multiple times (eg. with `package:` URIs and also filesystem paths) then they'll work fine. Dart Code used to do that but it caused other issues (I'm trying to hard to breakpoint resolution, and having two VM breakpoints for one VS Code breakpoint makes that harder) so we removed it once the changes where made that meant it shouldn't be required.

That said, there's a comment on there from only 3 days ago, so I wouldn't call that no activity :-)

**Comment by mezoni** [comment]
Many breakpoints do not work. Regardless of kinds of directives of import.
:-(
Everything works as if there are no these breakpoints.

**Comment by mezoni** [comment]
>> That said, there's a comment on there from only 3 days ago, so I wouldn't call that no activity :-)
Yeah...
It seems that developers do not care about your extension.
I remember they had an editor based on the Eclipse platform. At first, the editor was very good, until Dart developers handed it over to a person who honestly admitted that he was a student and didn’t care much about this development. Then its development was abandoned.

**Comment by mezoni** [comment]
The same problem in JetBrains IDE.
Most of breakpoints does not works.

**Comment by DanTup** [comment]
> Many breakpoints do not work. Regardless of kinds of directives of import.

If you have an issue that's not related to the type of import, please open a new issue with a log (capture it using **Dart: Capture Logs** and ticking just the **Debugger* option) and we'll see what we can find.

> It seems that developers do not care about your extension.

I'm not sure how you came to that conclusion in response to me saying the issue had been updated within the last few days, but it's certainly not true. I'm currently working full-time on this extension (and supporting tools) thanks to the Dart & Flutter teams!

> I remember they had an editor...

I don't know the history of this, but let's try to keep this issue tracker focused on making the VS Code extensions better :-)

**Comment by mezoni** [comment]
>> I'm currently working full-time on this extension (and supporting tools) thanks to the Dart & Flutter teams!

I know. Thank you for your hard work!
But unfortunately I temporarily switched to JetBrains IDE.
I will be very happy to switch back to VS Code when debugging will work as it should work. Unfortunately currently it does works partially.
P.S.
Possibly the problems are not in the VM communication protocol.
Maybe it should be at least fluently acquainted with the algorithms of other developers? For example with the code for the plugin for JetBrains IDE?
I do not have the slightest desire to downplay the importance of your product, but I just would like you to pay attention to this, because this issue is considered closed
At the all, in the end, not the implementation is important, but the end result.

**Comment by DanTup** [comment]
> I will be very happy to switch back to VS Code when debugging will work as it should work.

I'm more happy to look at the issue, but I need some information. If you can reproduce missed breakpoints that are not the issue mentioned above, please open a new issue with an Observatory log.


> Maybe it should be at least fluently acquainted with the algorithms of other developers? For example with the code for the plugin for JetBrains IDE?

I'm not sure what this means. I know of a difference between the VS Code debugger and IntelliJ one (sending breakpoints multiple times with different path formats), but I deliberately changed VS Code because it should no longer be required. If it turns out it is, I would prefer to raise that with the VM to try and have it fixed there rather than putting the hacks back in here (it complicates several things, including breakpoint resolution, which I've been trying to fix for over a year).

> I just would like you to pay attention to this, because this issue is considered closed

The issue described in the title was moved to the SDK tracker and this one closed as upstream. Once it's fixed in the SDK, Dart Code will get the fix automatically. A lot of issues opened here turn out to be issues that need to be fixed in the SDK(s) and keeping them all open here just makes it harder to know what issues can be progressed. If an issue requires some work in Dart Code once it's fixed in the SDK it remains open with the **Blocked by Dart / Flutter** label. If there's nothing to do in Dart Code, it gets closed with **Upstream in Dart / Flutter** (same goes for VS Code).

AFAIC, the issue appears to be being progressed - the SDK issue was marked as blocked and another issue opened. Developers are actively looking at this issue (at least, the one described in the title).


> Many breakpoints do not work. Regardless of kinds of directives of import.

Please do open an issue with more details about this. I can't investigate it given only that note. Please include an Observatory log. I *hate* having bugs about missed breakpoints and would very much like to get to the bottom of them.

**Comment by mezoni** [comment]
>> If you have an issue that's not related to the type of import, please open a new issue with a log (capture it using **Dart: Capture Logs** and ticking just the *_Debugger_ option) and we'll see what we can find.

In such log I found the following records:
- Breakpoint was registered as expected via the package relative path
- No events was send by VM about the pausing on this break points

So I can make conclusion that VM was never send back to your client the event about that this breakpoint was reached in the isolate execution process.

**Comment by DanTup** [comment]
@mezoni If you can open a new issue with the log (or email it to logs @ dartcode.org if you don't want to post it publicly) I can take a look. There are many things that can go wrong that would be easier to diagnose by seeing the log.

Alternatively, if you can make a small repro project you can share that shows the issue, I can try to repro here. But again, please open a new issue for it - it becomes confusing to manage multiple bugs in a single issue (because milestones, labels, etc. may diverge). Thanks!

**Comment by mezoni** [comment]
OK. When I have free time, I will create a small project and give it to you for analysis.
Thank you for your support.

**Comment by DanTup** [comment]
Thanks! A log may be just as useful if it's easier for you - I might be able to figure out the issue from what's in there.

**Comment by mezoni** [comment]
>> It's possible that if they send the breakpoints multiple times (eg. with `package:` URIs and also filesystem paths) then they'll work fine. Dart Code used to do that but it caused other issues (I'm trying to hard to breakpoint resolution, and having two VM breakpoints for one VS Code breakpoint makes that harder) so we removed it once the changes where made that meant it shouldn't be required.

Some comments to your words (the links to workable implementation).
Yes, to many additional work. But who can say that this additional work is useless and unneeded?

https://github.com/JetBrains/intellij-plugins/blob/4481281dcdeee5db6cf6909ce302021716b0f0a7/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java#L53

https://github.com/JetBrains/intellij-plugins/blob/4481281dcdeee5db6cf6909ce302021716b0f0a7/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceBreakpointHandler.java#L71

https://github.com/JetBrains/intellij-plugins/blob/4481281dcdeee5db6cf6909ce302021716b0f0a7/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceBreakpointHandler.java#L104

And aslo.


https://github.com/JetBrains/intellij-plugins/blob/5fb99c284ae67a4895da6120a349b270c9149699/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceDebugProcess.java#L453

https://github.com/JetBrains/intellij-plugins/blob/4481281dcdeee5db6cf6909ce302021716b0f0a7/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/frame/DartVmServiceEvaluator.java#L169

https://github.com/JetBrains/intellij-plugins/blob/4481281dcdeee5db6cf6909ce302021716b0f0a7/Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/VmServiceWrapper.java#L296

Please look at the whole body of the functions (implementation) but not to only at the signle line of the source code (provided by link).
It seems to me that they have also done no small work and are not going to give it up at the moment.
The logic of work used is not devoid of meaning, this implementation works, it is self-sufficient and it is fairly simple to understand in work.

**Comment by DanTup** [comment]
Unfortunately it's not as straight forward as copying the implementation from IntelliJ. VS Code has many quirks around breakpoints - I've many length discussions in their issue tracker trying to make this work (see https://github.com/Microsoft/vscode/issues/59984, https://github.com/Microsoft/vscode/issues/62264, https://github.com/Microsoft/vscode/issues/65993).

I have made progress with this, but it's still blocked on that last issue. It also required not sending multiple breakpoints to the VM for a single breakpoint in VS Code (which is what we used to do, and avoids the issue described here). The intention of the VM is to handle breakpoints predictably, so if that's not working, it makes sense to improve the VM rather than put the hacks back into VS Code.

If that doesn't happen in a reasonable timeframe, I would consider handling in VS Code, but first I need an actual repro to confirm the bug. The only case I've been able to confirm it not working is the one titled above, and a fix is in-progress.

If you're able to provide me a log file or repro of other bugs not covered by that issue, I will happily investigate, but I can't do so without more information.

**Comment by mezoni** [comment]
At which email send the project?

**Comment by mezoni** [comment]
I sent it to logs@dartcode.org.

**Comment by DanTup** [comment]
@mezoni Thanks, got it - though I see what's happened now - I'll add comments to https://github.com/Dart-Code/Dart-Code/issues/1489
